### PR TITLE
Fu 108 course edition redux thunk

### PR DESCRIPTION
**Pull request by Nikolay1703**

### CODE REVIEW

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 1]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
-        setTimeout (() => updateCourseThemes(), 200);
-    }
-
-    const deleteThemeCourse = (newThemeCourse: NewThemeCourse) => {
-        let url = 'Course/' + newThemeCourse.idCourse + '/theme/' + newThemeCourse.idTheme;
-        sendDeleteRequestNoResponse(url);
-        setTimeout (() => updateCourseThemes(), 200);
-    }
+    let nameThemesCourse: string[] = [];
+    let idCourse = Number(id);
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:28]: const idCourse = +id;

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 2]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
-        setNameThemes(nameThemesCourse);
-    }
-    
     useEffect(() => {
-        getAllThemes();
-        updateCourseThemes();
+        dispatch(getThemes());
+        dispatch(getCourseById(idCourse));
+        closeProgramCourse();
+        closeMaterialsCourse();
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:34]: closeProgramCourse();
closeMaterialsCourse();

are not needed here

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 3]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
     }, []);
 
-        
+    useEffect(() => {
+        checkThemes();
+    }, [pageState.course.themes]);
+
+    const [searchTurn, setSearchTurn] = useState('');
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:41]: rename to searchWord please

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 4]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
+            let newTheme: NewThemeCourse = {idCourse: idCourse, idTheme: theme.id};
+            dispatch(addThemeInCourse(newTheme));
+            openProgramCourse();
         }
     }
 
     const checkTheThemeInTheCourse = (theme: Themes): number => {
         let count = 0;
-        for (let item of themesCourse as Themes[]) {
+        for (let item of pageState.course.themes) {
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:53]: it's better to use something like
const theme = pageState.course.themes.find(t => t.id === theme.id)

Comment (Negodyaj) [reply src/components/courses-page/course-edition/CourseEdition.tsx:53]: return theme === undefined

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 5]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
                 count++;
                 break;
             }
         }
         return count;
+    } 
+
+    const checkThemes = () => {
+        pageState.course.themes?.map((theme) => (
+            nameThemesCourse.push(theme.name)
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:64]: it's better to rename to "themesInCourse"

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 6]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
-import { isCourse } from '../../../services/type-guards/course';
-
-interface NewThemeCourse {
+import { useParams } from 'react-router-dom';
+import { useDispatch, useSelector } from 'react-redux';
+import { IRootState } from '../../../store';
+import { addThemeInCourse, deleteThemeCourse, getCourseById, getThemes } from '../../../store/course-edition/thunk';
+import { setNameAllThemesInCourse, setChangeDisplayingButtonOpenProgramCourse, setChangeDisplayingButtonOpenMaterialsCourse, setChangeDisplayingButtonCloseProgramCourse, setChangeDisplayingButtonCloseMaterialsCourse } from '../../../store/course-edition/action-creators';
+
+export interface NewThemeCourse {
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:12]: please rename to CourseTheme

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 7]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
+        pageState.course.themes?.map((theme) => (
+            nameThemesCourse.push(theme.name)
+        ))
+        dispatch(setNameAllThemesInCourse(nameThemesCourse));
     }
 
     const deleteThemeFromCourse = (theme: Themes) => {
-        let newTheme: NewThemeCourse = {idCourse: indexCourse, idTheme: theme.id};
-        deleteThemeCourse(newTheme);
+        let newTheme: NewThemeCourse = {idCourse: idCourse, idTheme: theme.id};
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:70]: const courseTheme: CourseTheme

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 8]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
+        dispatch(deleteThemeCourse(newTheme));
     } 
 
     const searchFromTheme = (str: string) => {
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:74]: const searchInThemes = ...

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 9]
```diff
@@ -3,127 +3,97 @@ import { useEffect, useState } from 'react';
...
         setSearchTurn(str);
     }
 
     const openProgramCourse = () => {
-        setChangeDisplayingButtonOpenProgramCourse(!changeDisplayingButtonOpenProgramCourse);
-        isClassOnProgramCourse === ' unvisible' ? setIsClassOnProgramCourse(' visible') : setIsClassOnProgramCourse(' unvisible');
+        dispatch(setChangeDisplayingButtonOpenProgramCourse());
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:79]: 1. openSomething()  (show...)
2. closeSomething()  (hide...)

toggleSomething:
    propertyToChange: !state.propertyToChange

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 10]
```diff
@@ -135,7 +105,7 @@ function CourseEdition(props: CourseEditionProps) {
                 <div className="new-themes-container">
                     <SearchComponent funcSearch={searchFromTheme}/>
                     {
-                        allThemes?.filter((item) => {
+                        pageState.themes?.filter((item) => {
                             if (item.name.toLowerCase().includes(searchTurn.toLowerCase())) {
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:109]: pageState.themes?.filter(item => item.name.toLowerCase().includes(searchTurn.toLowerCase()))

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 11]
```diff
@@ -146,7 +116,7 @@ function CourseEdition(props: CourseEditionProps) {
                                 <div className="new-theme-add">
                                     <button onClick={() => addNewThemeInProgramCourse(item)} className="button-add-theme">
                                     {
-                                        nameThemes.includes(item.name) ? <FontAwesomeIcon icon="check" /> : <FontAwesomeIcon icon="plus" />
+                                        pageState.nameThemesCourse.includes(item.name) ? <FontAwesomeIcon icon="check" /> : <FontAwesomeIcon icon="plus" />
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:119]: it's better to use item.id

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 12]
```diff
@@ -157,17 +127,21 @@ function CourseEdition(props: CourseEditionProps) {
...
             <div className="current-course-container">
                 <div className="program-current-course-container">
-                    <div onClick={openProgramCourse}  className="program-course-header">
-                        <button onClick={openProgramCourse} className="program-course-header-button-open">
-                            {
-                                changeDisplayingButtonOpenProgramCourse ? <FontAwesomeIcon icon="angle-down" /> : <FontAwesomeIcon icon="angle-up" />
+                        <div className="program-course-header">
+                            { pageState.isDisplayingButtonOpenProgramCourse ?
+                                <button onClick={closeProgramCourse} className="program-course-header-button-open">
+                                    <FontAwesomeIcon icon="angle-down" />
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:133]: use condition inside this attribute

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 13]
```diff
@@ -157,17 +127,21 @@ function CourseEdition(props: CourseEditionProps) {
...
                             }
-                        </button>
-                        <div className="program-course-header-text">Программа курса</div>
-                    </div>
-                    <div className={"program-course" + isClassOnProgramCourse}>
-                        {
-                            themesCourse?.map((theme) => (
+                            <div className="program-course-header-text">Программа курса</div>
+                        </div>
+                    <div className={"program-course"}>
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:142]: "{ }" are not needed

File: src/components/courses-page/course-edition/CourseEdition.tsx [hunk 14]
```diff
@@ -181,12 +155,16 @@ function CourseEdition(props: CourseEditionProps) {
...
-                    <div onClick={openMaterialsCourse} className={"materials-course-header" + isClassOnMaterialTheme}>
-                        <button onClick={openMaterialsCourse} className="materials-course-header-button-open">
-                            {
-                                changeDisplayingButtonOpenMaterialsCourse ? <FontAwesomeIcon icon="angle-down" /> : <FontAwesomeIcon icon="angle-up" />
-                            }
-                        </button>
+                    <div className={"materials-course-header"}>
+                        { pageState.isDisplayingButtonOpenMaterialsCourse ?
+                            <button onClick={closeMaterialsCourse} className="materials-course-header-button-open">
+                                <FontAwesomeIcon icon="angle-down" />
```
Comment (Negodyaj) [inline src/components/courses-page/course-edition/CourseEdition.tsx:161]: use condition inside this attribute

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 1 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:28]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 3 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:41]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 4 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:53]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 5 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:64]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 10 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:109]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 13 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:142]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 12 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:133]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 14 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:161]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 6 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:12]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 7 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:70]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 8 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:74]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 9 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:79]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 11 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:119]: fixed

File: src/components/courses-page/course-edition/CourseEdition.tsx (hunk 2 above)
Comment (Nikolay1703) [reply src/components/courses-page/course-edition/CourseEdition.tsx:34]: fixed

**Pull request merged**